mypy/checker.py

```diff
@@ -1292,8 +1292,9 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                 self.check_indexed_assignment(index_lvalue, rvalue, lvalue)
 
             if inferred:
-                self.infer_variable_type(inferred, lvalue, self.expr_checker.accept(rvalue),
-                                         rvalue)
+                is_cast = isinstance(rvalue, CallExpr) and rvalue.is_cast()
+                init_type = self.expr_checker.accept(rvalue, always_allow_any=is_cast)
+                self.infer_variable_type(inferred, lvalue, init_type, rvalue)
```

Comment:
For the following case: 
```py
m = cast(Any, Foo().g)
```

After splitting up the test cases like you suggested, I realized that that case doesn't work.